### PR TITLE
Fix: Prevent creating folders with forbidden names

### LINKED ISSUE #9522: Bug: Files is allowing forbidden folder names (CON, NUL, etc)
### Description

It seems that Files doesn't prevent the user from creating files or folders with forbidden names that regular File Explorer would otherwise prevent.

It should prevent the user from creating files or folders named (in either lowercase or uppercase):

CON
PRN
AUX
NUL
COM1, COM2, COM3, COM4, COM5, COM6, COM7, COM8, COM9, COM0
LPT1, LPT2, LPT3, LPT4, LPT5, LPT6, LPT7, LPT8, LPT9, LPT0

con01
con02

### Steps To Reproduce

1. Open Files
2. Go to a place where you can freely create folders or files
3. Create a folder or a file with a forbidden name (list in the description)
4. Open file explorer and create a folder or a file with the same name
5. File explorer will not allow the creation of a folder or a file with that name

### Steps to fix

- Display text when forbidden folders names are entered into the new folder modal
- Display error content dialog when users try renaming a folder to con

### Files Version

2.3.3.0

### Windows Version

Windows 11 21H2 22000.778

### Relevant Assets/Logs

n/a

### PR DESCRIPTION
**Pull request by hecksmosis**
**Resolved / Related Issues**
Items resolved / related issues by this PR.
- Closes #9522 

**Validation**
How did you test these changes?
- [x] Built and ran the app

### CODE REVIEW

File: src/Files.App/Filesystem/FilesystemOperations/FilesystemOperations.cs [hunk 1]
```diff
@@ -53,7 +53,7 @@ public async Task<(IStorageHistory, IStorageItem)> CreateAsync(IStorageItemWithP
 			FileSystemProgress fsProgress = new(progress, true, FileSystemStatusCode.InProgress, 1);
 			fsProgress.Report();
 
-			try
+				try
```
Comment (yaira2) [inline src/Files.App/Filesystem/FilesystemOperations/FilesystemOperations.cs:56]: ?

File: src/Files.App/Filesystem/FilesystemOperations/Helpers/FilesystemHelpers.cs [hunk 2]
```diff
@@ -96,6 +96,13 @@ public async Task<(ReturnResult, IStorageItem)> CreateAsync(IStorageItemWithPath
 			var progress = new Progress<FileSystemProgress>();
 			progress.ProgressChanged += (s, e) => returnStatus = returnStatus < ReturnResult.Failed ? e.Status!.Value.ToStatus() : returnStatus;
 
+			if (string.IsNullOrWhiteSpace(source.Name)
+				|| ContainsRestrictedCharacters(source.Name)
+				|| ContainsRestrictedFileName(source.Name))
+			{
+				return (ReturnResult.Failed, null);
```
Comment (yaira2) [inline src/Files.App/Filesystem/FilesystemOperations/Helpers/FilesystemHelpers.cs:103]: Does this display a dialog?

Comment (hecksmosis) [reply src/Files.App/Filesystem/FilesystemOperations/Helpers/FilesystemHelpers.cs:102]: No, but in CreateAsync no error displays a dialog

Comment (hecksmosis) [reply src/Files.App/Filesystem/FilesystemOperations/Helpers/FilesystemHelpers.cs:104]: Should I change it? @yair100

Comment (yaira2) [reply src/Files.App/Filesystem/FilesystemOperations/Helpers/FilesystemHelpers.cs:102]: What does the error dialog look like?

Comment (hecksmosis) [reply src/Files.App/Filesystem/FilesystemOperations/Helpers/FilesystemHelpers.cs:102]: The error dialog could be something like this, if we use existing strings already in resources.resw
image

Comment (yaira2) [reply src/Files.App/Filesystem/FilesystemOperations/Helpers/FilesystemHelpers.cs:102]: Strings would have to be modified but this is the right idea.

Comment (hecksmosis) [reply src/Files.App/Filesystem/FilesystemOperations/Helpers/FilesystemHelpers.cs:102]: Ok, I'll fix it now

Comment (hecksmosis) [reply src/Files.App/Filesystem/FilesystemOperations/Helpers/FilesystemHelpers.cs:106]: Done

**Comment by gave92** [comment]
Why are not just checking for invalid names in the create file dialog?

**Comment by yaira2** [comment]
> Why are not just checking for invalid names in the create file dialog?

Need to cover rename to as well.

**Comment by hecksmosis** [comment]
Rename already has the check

**Comment by hecksmosis** [comment]
If we cover this in the dialog we would have to check in multiple places, wouldn't we?

**Comment by yaira2** [comment]
> Rename already has the check

What dialog is displayed there? We should reuse as much code as possible.

**Comment by hecksmosis** [comment]
I've noticed there isn't actually a check for rename. (I've added one)

File: src/Files.App/Filesystem/FilesystemOperations/Helpers/FilesystemHelpers.cs [hunk 3]
```diff
@@ -538,6 +548,16 @@ public async Task<ReturnResult> RenameAsync(IStorageItemWithPath source, string
...
 
+			if (string.IsNullOrWhiteSpace(newName)
+				|| ContainsRestrictedCharacters(newName)
+				|| ContainsRestrictedFileName(newName))
+			{
+				await DialogDisplayHelper.ShowDialogAsync(
+					"ErrorDialogThisActionCannotBeDone".GetLocalizedResource(),
+					"ErrorDialogNameNotAllowed".GetLocalizedResource());
+				return ReturnResult.Failed;
+			}
```
Comment (yaira2) [inline src/Files.App/Filesystem/FilesystemOperations/Helpers/FilesystemHelpers.cs:559]: Is it worth creating a common method for this code?

Comment (hecksmosis) [reply src/Files.App/Filesystem/FilesystemOperations/Helpers/FilesystemHelpers.cs:559]: The return types change, so I think we should leave it as is

File: src/Files.App/Strings/en-US/Resources.resw [hunk 4]
```diff
@@ -2877,4 +2877,7 @@
   <data name="ReleaseNotes" xml:space="preserve">
     <value>Release Notes</value>
   </data>
+  <data name="ErrorDialogNameNotAllowed" xml:space="preserve">
+    <value>The item name specfied is invalid</value>
```
Comment (yaira2) [inline src/Files.App/Strings/en-US/Resources.resw:2881]:
```suggestion
    <value>The specified item name is invalid</value>
```

File: src/Files.App/Filesystem/FilesystemOperations/Helpers/FilesystemHelpers.cs [hunk 5]
```diff
@@ -538,6 +548,16 @@ public async Task<ReturnResult> RenameAsync(IStorageItemWithPath source, string
 			var progress = new Progress<FileSystemProgress>();
 			progress.ProgressChanged += (s, e) => returnStatus = returnStatus < ReturnResult.Failed ? e.Status!.Value.ToStatus() : returnStatus;
 
+			if (string.IsNullOrWhiteSpace(newName)
+				|| ContainsRestrictedCharacters(newName)
+				|| ContainsRestrictedFileName(newName))
```
Comment (QuaintMako) [inline src/Files.App/Filesystem/FilesystemOperations/Helpers/FilesystemHelpers.cs:553]: This condition is tested twice (at least) in the file. It should be extracted inside a method called `IsSuitedForFileName(string name)` or something akin.

File: src/Files.App/Filesystem/FilesystemOperations/Helpers/FilesystemHelpers.cs [hunk 6]
```diff
@@ -538,6 +548,16 @@ public async Task<ReturnResult> RenameAsync(IStorageItemWithPath source, string
...
+			{
+				await DialogDisplayHelper.ShowDialogAsync(
+					"ErrorDialogThisActionCannotBeDone".GetLocalizedResource(),
+					"ErrorDialogNameNotAllowed".GetLocalizedResource());
+				return ReturnResult.Failed;
```
Comment (QuaintMako) [inline src/Files.App/Filesystem/FilesystemOperations/Helpers/FilesystemHelpers.cs:558]: This block is almost the same as l.103 to l.107 aside of the return time. Would it be possible to fuse them? I'm not sure how, it's not huge. Except if you find an idea right away, do not spend too much on it.

Comment (hecksmosis) [reply src/Files.App/Filesystem/FilesystemOperations/Helpers/FilesystemHelpers.cs:558]: I don't think there is an easy way to merge them without modifying changing a lot of code

Comment (QuaintMako) [reply src/Files.App/Filesystem/FilesystemOperations/Helpers/FilesystemHelpers.cs:558]: Agreeing with you on this one, let's not overdo it.

**Review by QuaintMako — DISMISSED** [review]

**Review by QuaintMako — APPROVED** [review]

**Comment by yaira2** [comment]
@hecksmosis thank you

**Pull request merged**